### DISCUSSION TITLE
Reading emails from Gmail using IMAP in C# is slow

### DISCUSSION THREAD

**Discussion by bhumeshdeekonda** [jstedfast/MailKit#1820] [Q&A]
**Hi everyone,**

I’m having trouble with retrieving emails using IMAP in C#. Here’s what I’m trying to do:

- Connect to Gmail using IMAP.
- Fetch recent emails efficiently.

I’m currently using the MailKit library and the following code:

```csharp
using (var client = new ImapClient())
{
    client.Connect("imap.gmail.com", 993, true);
    client.Authenticate("your-email@gmail.com", "your-password");

    var inbox = client.Inbox;
    inbox.Open(FolderAccess.ReadOnly);

    // Use a search to only fetch recent emails
    var query = SearchQuery.Recent;
    var uids = inbox.Search(query);

    foreach (var uid in uids)
    {
        var message = inbox.GetMessage(uid);
    }

    client.Disconnect(true);
}
```

However, it’s taking a long time to fetch emails, and I’m not sure if I’m doing something wrong.

I’ve already tried limiting the number of messages retrieved.
My internet connection seems fine.
Any advice on optimizing this would be greatly appreciated!

Thank you!

**Comment by jstedfast** [comment]
I guess my first question would be to understand your expectations. When you say that it's slow, compared to what?

If your expectation is that it should be as fast as any IMAP client populates a message list that you can then select to read a particular email, then that's because you are comparing apples to oranges.

Outlook and Thunderbird, for example, do not download every message in order to populate that message list. They only fetch summary information about the messages.

You can fetch summary information about the messages, too:

```csharp
var summaries = client.Inbox.Fetch (uids, MessageSummaryItems.Envelope | MessageSummaryItems.Flags);
```

The problem you are facing is that messages themselves can be very large which is slow to download. This is made worse by the fact that you've got a loop that calls `GetMessage()`. Each call to `GetMessage()` requires a new command be sent to the server and the client must then wait for a response. The further the IMAP server is from the client, the more latency there will be (compounded if the server is busy with other client requests as well).

You'll notice that most of MailKit's ImapClient/ImapFolder APIs allow batch requests in order to alleviate this issue, but it does not have a `GetMessages()` API to allow batch requesting multiple messages. That said, MailKit *does* provide a `GetStreams()` method that can be used if you truly need to batch request a bunch of messages. However, the [GetStreams()](https://mimekit.net/docs/html/M_MailKit_Net_Imap_ImapFolder_GetStreams.htm) method will require you to pass a callback delegate to the method that can be used to process each raw message stream.

**Reply by ssousa97** [reply]
Would i be able to do something like:

```
var messages = await sentFolder.FetchAsync(results.UniqueIds, MessageSummaryItems.BodyStructure | MessageSummaryItems.Envelope);
var task = messages
    .Where(msg => Regex.IsMatch(msg.Envelope.Subject, regex))
    .Select(msg => sentFolder.GetBodyPartAsync(msg.UniqueId, msg.TextBody));
var completed = await Task.WhenAll(task);

foreach (var msg in completed) {
    Console.WriteLine(msg); 
}
```
I actually tried but got the SyncRoot error and i don't know how to do it (if its even possible)
Also, i didn't found this GetStreams() method, only GetStream

Edit: Nvm, just discovered how to use the GetStreams, i had to cast to (ImapFolder) instead of (IMailFolder) from the GetFolder method

**Reply by jstedfast** [reply]
> I actually tried but got the SyncRoot error and i don't know how to do it (if its even possible)

The issue you ran into was that your code attempts to start a whole bunch of "GetBodyPartAsync" jobs that will all be in a running state at the same time. Keep in mind that the client can only handle 1 request at a time, so this will cause problems.

The way to fix your above code would be something more like this:

```csharp
var messages = await sentFolder.FetchAsync(results.UniqueIds, MessageSummaryItems.BodyStructure | MessageSummaryItems.Envelope);
var matches = messages.Where(msg => Regex.IsMatch(msg.Envelope.Subject, regex));

foreach (var match in matches) {
    var msg = await sentFolder.GetBodyPartAsync(match.UniqueId, match.TextBody));
    Console.WriteLine(msg); 
}
```
